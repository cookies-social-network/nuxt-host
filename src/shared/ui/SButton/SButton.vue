<template>
  <component :is="props.tag" :disabled="props.disabled || props.loading" :class="classes" v-bind="$attrs" v-on="$attrs">
    <span v-if="slots.prefix" class="s-button__prefix">
      <slot name="prefix" />
    </span>
    <!-- пустой класс нужен для управления loading -->
    <span class="s-button__content">
      <slot />
    </span>
    <span v-if="slots.suffix" class="s-button__suffix">
      <slot name="suffix" />
    </span>
  </component>
</template>

<script lang="ts" setup>
type Props = {
  tag?: 'button' | 'a'
  size?: 'md' | 'sm'
  loading?: boolean
  disabled?: boolean
  // DEBT: реализовать в одном стиле линки. Либо через класс .link, либо через компонент s-button. Как вариант можно завести переменные чтобы прокинуть в оба варианта
  appearance?: 'primary' | 'secondary' | 'tertiary' | 'link'
  onlyIcon?: boolean
}

const props = withDefaults(defineProps<Props>(), {
  tag: 'button',
  appearance: 'primary',
  loading: false,
  disabled: false,
  size: 'md',
  onlyIcon: false,
})

const slots = useSlots()

const classes = computed(() => {
  return [
    's-button',
    `s-button--${props.appearance}`,
    `s-button--${props.size}`,
    {
      's-button--prefix': slots.prefix,
      's-button--suffix': slots.suffix,
      's-button--loading': props.loading,
      's-button--only-icon': props.onlyIcon,
    },
  ]
})
</script>

<style lang="scss" scoped>
.s-button {
  --color-white: #fff;

  color: var(--color-white);
}
</style>
