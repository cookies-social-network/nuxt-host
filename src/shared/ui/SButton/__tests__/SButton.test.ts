import { mount } from '@vue/test-utils'
import { describe, test, expect } from 'vitest'

import SButton from '../SButton.vue'

describe('Testing SButton', async () => {
  test('Mounting', () => {
    const button = mount(SButton, {
      props: {
        tag: 'button',
      },
      slots: {
        default: 'Привет',
      },
    })

    expect(button.element).toMatchSnapshot()
  })
})
