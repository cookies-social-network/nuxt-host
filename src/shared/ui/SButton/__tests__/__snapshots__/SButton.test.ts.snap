// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing SButton > Mounting 1`] = `
<button
  class="s-button s-button--primary s-button--md"
  data-v-aa378b68=""
>
  <!--v-if-->
  <!-- пустой класс нужен для управления loading -->
  <span
    class="s-button__content"
    data-v-aa378b68=""
  >
    
    Привет
    
  </span>
  <!--v-if-->
</button>
`;
